File: go/problem_13.go
package main

import "fmt"

// longestSubstringKDistinct returns the length of the longest substring
// containing at most k distinct characters.
// Uses a sliding window with a character-frequency map.
// This runs in O(n) time and O(k) space.
func longestSubstringKDistinct(s string, k int) int {
	freq := make(map[rune]int)
	left, best := 0, 0
	runes := []rune(s)

	for right := 0; right < len(runes); right++ {
		freq[runes[right]]++ // expand window to the right

		// shrink from left until we have at most k distinct chars
		for len(freq) > k {
			freq[runes[left]]--
			if freq[runes[left]] == 0 {
				delete(freq, runes[left])
			}
			left++
		}

		if right-left+1 > best {
			best = right - left + 1
		}
	}
	return best
}

func problem13() {
	fmt.Println("Problem 13:")
	if longestSubstringKDistinct("abcba", 2) != 3 { panic("assertion failed") }
	if longestSubstringKDistinct("aa", 1)    != 2 { panic("assertion failed") }
	if longestSubstringKDistinct("abc", 3)   != 3 { panic("assertion failed") }
	if longestSubstringKDistinct("", 2)      != 0 { panic("assertion failed") }
	fmt.Println("All Tests Passed")
}

File: rust/problem_13.rs
use std::collections::HashMap;

// longest_substring_k_distinct returns the length of the longest substring
// containing at most k distinct characters.
// Uses a sliding window. This runs in O(n) time and O(k) space.
pub fn longest_substring_k_distinct(s: &str, k: usize) -> usize {
    let chars: Vec<char> = s.chars().collect();
    let mut freq: HashMap<char, usize> = HashMap::new();
    let mut left = 0usize;
    let mut best = 0usize;

    for right in 0..chars.len() {
        *freq.entry(chars[right]).or_insert(0) += 1; // expand window

        // shrink from left until at most k distinct chars remain
        while freq.len() > k {
            let cnt = freq.get_mut(&chars[left]).unwrap();
            *cnt -= 1;
            if *cnt == 0 { freq.remove(&chars[left]); }
            left += 1;
        }

        best = best.max(right - left + 1);
    }
    best
}

pub fn problem13() {
    println!("Problem 13:");
    assert_eq!(longest_substring_k_distinct("abcba", 2), 3, "assertion failed");
    assert_eq!(longest_substring_k_distinct("aa", 1),    2, "assertion failed");
    assert_eq!(longest_substring_k_distinct("abc", 3),   3, "assertion failed");
    assert_eq!(longest_substring_k_distinct("", 2),      0, "assertion failed");
    println!("All Tests Passed");
}
